src/coreclr/jit/codegencommon.cpp

```diff
@@ -9442,3 +9442,30 @@ bool CodeGen::genCanOmitNormalizationForBswap16(GenTree* tree)
 
     return (cast->gtCastType == TYP_USHORT) || (cast->gtCastType == TYP_SHORT);
 }
+
+//----------------------------------------------------------------------
+// genCodeForReuseVal: Generate code for a node marked with re-using a register.
+//
+// Arguments:
+//   tree - The node marked with re-using a register
+//
+// Remarks:
+//   Generates nothing, except for when the node is a CNS_INT(0) where
+//   we will define a new label to propagate GC info. We want to do this
+//   because if the node is a CNS_INT(0) and is re-using a register,
+//   that register could have been used for a CNS_INT(ref null) that is GC
+//   tracked.
+//
+void CodeGen::genCodeForReuseVal(GenTree* treeNode)
+{
+    assert(treeNode->IsReuseRegVal());
+
+    // For now, this is only used for constant nodes.
+    assert(treeNode->OperIs(GT_CNS_INT, GT_CNS_DBL, GT_CNS_VEC));
+    JITDUMP("  TreeNode is marked ReuseReg\n");
+
+    if (treeNode->IsIntegralConst(0))
+    {
+        genDefineTempLabel(genCreateTempLabel());
```

Comment:
This appears to be the only way to update GC info properly.

Reply:
I would think the right fix is something like
```C++
gcInfo.gcMarkRegPtrVal(tree->getRegNum(), tree->TypeGet());
```

Reply:
@AndyAyersMS In order for the state of `gcInfo` to be reflected in the actual emitting of instructions, it has to be propagated via instruction groups, which is what the creation of the label is doing. In this case, the state of `gcInfo` is already marked as a non-pointer, but that information wasn't propagated. `gcMarkRegPtrVal` doesn't propagate the info, it just sets the state.

Reply:
It does propagate it, but only during codegen. I suspect the emitter GC tracking doesn't handle a case like this where the gcness of a register can change without writing to the register.

In which case, you might consider introducing a fake instruction that just reinterprets the register with its new GC type and acts like a write; that might convey intention of changing GC type more clearly than forcing a label, which seems a bit fragile and obscure.

Or it might gum up other things. Hard to say.

Reply:
> you might consider introducing a fake instruction

Ah, I see Bruce had similar comments below.

I agree we can take this fix for now, but we should seriously consider something a bit more explicit. Ideally the code generator would not need to be aware of the peculiar aspects of emitter gc tracking.

Reply:
I agree that we should introduce a fake instruction, this is what @BruceForstall and I talked about. Seems Bruce beat me to making an [issue](https://github.com/dotnet/runtime/issues/84313) about it. :)